Acceptance test

* from time to time the acceptance tests fail with a timeout issue
* see https://drone.nextcloud.com/nextcloud/server/7415/32

@danxuliu could you have a look?

Log:

```
07:10:40.651 WARN - Exception thrown
org.openqa.selenium.StaleElementReferenceException: Element is no longer attached to the DOM
Command duration or timeout: 56 milliseconds
For documentation on this error, please visit: http://seleniumhq.org/exceptions/stale_element_reference.html
Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
System info: host: '3a30ab296074', ip: '172.17.0.10', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-31-generic', java.version: '1.8.0_91'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Capabilities [{applicationCacheEnabled=true, rotatable=false, handlesAlerts=true, databaseEnabled=true, version=47.0.1, platform=LINUX, nativeEvents=false, acceptSslCerts=true, webStorageEnabled=true, locationContextEnabled=true, browserName=firefox, takesScreenshot=true, javascriptEnabled=true, cssSelectorsEnabled=true}]
Session ID: 01a16c11-4572-49d2-8374-9ed291bfc1d7
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:678)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:327)
	at org.openqa.selenium.remote.RemoteWebElement.getAttribute(RemoteWebElement.java:150)
	at sun.reflect.GeneratedMethodAccessor11.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.openqa.selenium.support.events.EventFiringWebDriver$EventFiringWebElement$1.invoke(EventFiringWebDriver.java:335)
	at com.sun.proxy.$Proxy2.getAttribute(Unknown Source)
	at org.openqa.selenium.support.events.EventFiringWebDriver$EventFiringWebElement.getAttribute(EventFiringWebDriver.java:373)
	at sun.reflect.GeneratedMethodAccessor11.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.openqa.selenium.remote.server.KnownElements$1.invoke(KnownElements.java:64)
	at com.sun.proxy.$Proxy3.getAttribute(Unknown Source)
	at org.openqa.selenium.remote.server.handler.GetElementAttribute.call(GetElementAttribute.java:36)
	at org.openqa.selenium.remote.server.handler.GetElementAttribute.call(GetElementAttribute.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.openqa.selenium.remote.server.DefaultSession$1.run(DefaultSession.java:176)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.openqa.selenium.StaleElementReferenceException: Element is no longer attached to the DOM
For documentation on this error, please visit: http://seleniumhq.org/exceptions/stale_element_reference.html
Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
System info: host: '3a30ab296074', ip: '172.17.0.10', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-31-generic', java.version: '1.8.0_91'
Driver info: driver.version: unknown
	at <anonymous class>.fxdriver.cache.getElementAt(resource://fxdriver/modules/web-element-cache.js:9457)
	at <anonymous class>.Utils.getElementAt(file:///tmp/anonymous7680842731605908180webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:9039)
	at <anonymous class>.WebElement.getElementAttribute(file:///tmp/anonymous7680842731605908180webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12146)
	at <anonymous class>.DelayedCommand.prototype.executeInternal_/h(file:///tmp/anonymous7680842731605908180webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12661)
	at <anonymous class>.DelayedCommand.prototype.executeInternal_(file:///tmp/anonymous7680842731605908180webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12666)
	at <anonymous class>.DelayedCommand.prototype.execute/<(file:///tmp/anonymous7680842731605908180webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12608)
07:10:40.658 WARN - Exception: Element is no longer attached to the DOM
For documentation on this error, please visit: http://seleniumhq.org/exceptions/stale_element_reference.html
Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
System info: host: '3a30ab296074', ip: '172.17.0.10', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-31-generic', java.version: '1.8.0_91'
Driver info: driver.version: unknown
```

--- a/tests/acceptance/features/bootstrap/FilesAppContext.php
+++ b/tests/acceptance/features/bootstrap/FilesAppContext.php
@@ -106,7 +106,11 @@ class FilesAppContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	public static function shareLinkCheckbox() {
-		return Locator::forThe()->content("Share link")->descendantOf(self::currentSectionDetailsView())->
+		// forThe()->checkbox("Share link") can not be used here; that would
+		// return the checkbox itself, but the element that the user interacts
+		// with is the label.
+		return Locator::forThe()->xpath("//label[normalize-space() = 'Share link']")->
+				descendantOf(self::currentSectionDetailsView())->
 				describedAs("Share link checkbox in the details view in Files app");
 	}
 
@@ -122,7 +126,11 @@ class FilesAppContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	public static function passwordProtectCheckbox() {
-		return Locator::forThe()->content("Password protect")->descendantOf(self::currentSectionDetailsView())->
+		// forThe()->checkbox("Password protect") can not be used here; that
+		// would return the checkbox itself, but the element that the user
+		// interacts with is the label.
+		return Locator::forThe()->xpath("//label[normalize-space() = 'Password protect']")->
+				descendantOf(self::currentSectionDetailsView())->
 				describedAs("Password protect checkbox in the details view in Files app");
 	}
 
@@ -163,7 +171,7 @@ class FilesAppContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	public static function favoritedStateIconForFile($fileName) {
-		return Locator::forThe()->content("Favorited")->descendantOf(self::favoriteActionForFile($fileName))->
+		return Locator::forThe()->css(".icon-starred")->descendantOf(self::favoriteActionForFile($fileName))->
 				describedAs("Favorited state icon for file $fileName in Files app");
 	}
 
@@ -210,7 +218,8 @@ class FilesAppContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	private static function fileActionsMenuItemFor($itemText) {
-		return Locator::forThe()->content($itemText)->descendantOf(self::fileActionsMenu())->
+		return Locator::forThe()->xpath("//a[normalize-space() = '$itemText']")->
+				descendantOf(self::fileActionsMenu())->
 				describedAs($itemText . " item in file actions menu in Files app");
 	}
 

--- a/tests/acceptance/features/bootstrap/FilesSharingAppContext.php
+++ b/tests/acceptance/features/bootstrap/FilesSharingAppContext.php
@@ -47,7 +47,7 @@ class FilesSharingAppContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	public static function wrongPasswordMessage() {
-		return Locator::forThe()->content("The password is wrong. Try again.")->
+		return Locator::forThe()->xpath("//*[@class = 'warning' and normalize-space() = 'The password is wrong. Try again.']")->
 				describedAs("Wrong password message in Authenticate page");
 	}
 

--- a/tests/acceptance/features/bootstrap/LoginPageContext.php
+++ b/tests/acceptance/features/bootstrap/LoginPageContext.php
@@ -66,7 +66,7 @@ class LoginPageContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	public static function wrongPasswordMessage() {
-		return Locator::forThe()->content("Wrong password. Reset it?")->
+		return Locator::forThe()->xpath("//*[@class = 'warning' and normalize-space() = 'Wrong password. Reset it?']")->
 				describedAs("Wrong password message in Login page");
 	}
 

--- a/tests/acceptance/features/bootstrap/NotificationContext.php
+++ b/tests/acceptance/features/bootstrap/NotificationContext.php
@@ -31,7 +31,8 @@ class NotificationContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	public static function notificationMessage($message) {
-		return Locator::forThe()->content($message)->descendantOf(self::notificationContainer())->
+		return Locator::forThe()->xpath("//*[@class = 'row' and normalize-space() = '$message']")->
+				descendantOf(self::notificationContainer())->
 				describedAs("$message notification");
 	}
 

--- a/tests/acceptance/features/bootstrap/SettingsMenuContext.php
+++ b/tests/acceptance/features/bootstrap/SettingsMenuContext.php
@@ -61,7 +61,8 @@ class SettingsMenuContext implements Context, ActorAwareInterface {
 	 * @return Locator
 	 */
 	private static function menuItemFor($itemText) {
-		return Locator::forThe()->content($itemText)->descendantOf(self::settingsMenu())->
+		return Locator::forThe()->xpath("//a[normalize-space() = '$itemText']")->
+				descendantOf(self::settingsMenu())->
 				describedAs($itemText . " item in Settings menu");
 	}
 

--- a/tests/acceptance/features/core/Locator.php
+++ b/tests/acceptance/features/core/Locator.php
@@ -134,7 +134,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function id($value) {
-		return $this->customSelector("named", array("id", $value));
+		return $this->customSelector("named_exact", array("id", $value));
 	}
 
 	/**
@@ -142,7 +142,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function idOrName($value) {
-		return $this->customSelector("named", array("id_or_name", $value));
+		return $this->customSelector("named_exact", array("id_or_name", $value));
 	}
 
 	/**
@@ -150,7 +150,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function link($value) {
-		return $this->customSelector("named", array("link", $value));
+		return $this->customSelector("named_exact", array("link", $value));
 	}
 
 	/**
@@ -158,7 +158,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function button($value) {
-		return $this->customSelector("named", array("button", $value));
+		return $this->customSelector("named_exact", array("button", $value));
 	}
 
 	/**
@@ -166,7 +166,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function linkOrButton($value) {
-		return $this->customSelector("named", array("link_or_button", $value));
+		return $this->customSelector("named_exact", array("link_or_button", $value));
 	}
 
 	/**
@@ -174,7 +174,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function field($value) {
-		return $this->customSelector("named", array("field", $value));
+		return $this->customSelector("named_exact", array("field", $value));
 	}
 
 	/**
@@ -182,7 +182,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function selectField($value) {
-		return $this->customSelector("named", array("select", $value));
+		return $this->customSelector("named_exact", array("select", $value));
 	}
 
 	/**
@@ -190,7 +190,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function checkbox($value) {
-		return $this->customSelector("named", array("checkbox", $value));
+		return $this->customSelector("named_exact", array("checkbox", $value));
 	}
 
 	/**
@@ -198,7 +198,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function radioButton($value) {
-		return $this->customSelector("named", array("radio", $value));
+		return $this->customSelector("named_exact", array("radio", $value));
 	}
 
 	/**
@@ -206,7 +206,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function fileInput($value) {
-		return $this->customSelector("named", array("file", $value));
+		return $this->customSelector("named_exact", array("file", $value));
 	}
 
 	/**
@@ -214,7 +214,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function optionGroup($value) {
-		return $this->customSelector("named", array("optgroup", $value));
+		return $this->customSelector("named_exact", array("optgroup", $value));
 	}
 
 	/**
@@ -222,7 +222,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function option($value) {
-		return $this->customSelector("named", array("option", $value));
+		return $this->customSelector("named_exact", array("option", $value));
 	}
 
 	/**
@@ -230,7 +230,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function fieldSet($value) {
-		return $this->customSelector("named", array("fieldset", $value));
+		return $this->customSelector("named_exact", array("fieldset", $value));
 	}
 
 	/**
@@ -238,7 +238,7 @@ class LocatorBuilder {
 	 * @return LocatorBuilderSecondStep
 	 */
 	public function table($value) {
-		return $this->customSelector("named", array("table", $value));
+		return $this->customSelector("named_exact", array("table", $value));
 	}
 
 }

--- a/tests/acceptance/features/core/Locator.php
+++ b/tests/acceptance/features/core/Locator.php
@@ -173,14 +173,6 @@ class LocatorBuilder {
 	 * @param string $value
 	 * @return LocatorBuilderSecondStep
 	 */
-	public function content($value) {
-		return $this->customSelector("named", array("content", $value));
-	}
-
-	/**
-	 * @param string $value
-	 * @return LocatorBuilderSecondStep
-	 */
 	public function field($value) {
 		return $this->customSelector("named", array("field", $value));
 	}

